File: go/refactoring/parallel_change/sprout_change/sprout_change.go
package sprout_change

import "math"

type CartItem struct {
	ID       string
	Price    float64
	Qty      int
	Category string
}

type Region string

const (
	RegionUS Region = "US"
	RegionEU Region = "EU"
)

type TaxPolicy interface {
}

// CalculateTotal: regla existente: un único impuesto plano por región; los libros y la comida están exentos en la UE
func CalculateTotal(cart []CartItem, region Region) float64 {
	subtotal := 0.0
	for _, it := range cart {
		subtotal += it.Price * float64(it.Qty)
	}

	tax := 0.0
	if region == RegionUS {
		tax = subtotal * 0.07 // 7% plano
	} else if region == RegionEU {
		// exenciones ingenuas en línea
		taxable := 0.0
		for _, it := range cart {
			if it.Category != "books" && it.Category != "food" {
				taxable += it.Price * float64(it.Qty)
			}
		}
		tax = taxable * 0.2 // 20% plano solo sobre los ítems gravables
	}

	return RoundCurrency(subtotal + tax)
}

func RoundCurrency(amount float64) float64 {
	return math.Round(amount*100) / 100
}

// DemoSprout: uso de ejemplo, mantenido simple para estudiantes
func DemoSprout() float64 {
	cart := []CartItem{
		{ID: "p1", Price: 10, Qty: 2, Category: "general"},
		{ID: "b1", Price: 20, Qty: 1, Category: "books"},
	}
	return CalculateTotal(cart, RegionEU)
}

File: java/src/main/java/com/refactoring/parallelchange/sproutchange/SproutChange.java
package com.refactoring.parallelchange.sproutchange;

import java.util.List;

public class SproutChange {

    public static class CartItem {
        public final String id;
        public final double price;
        public final int qty;
        public final String category;

        public CartItem(String id, double price, int qty, String category) {
            this.id = id;
            this.price = price;
            this.qty = qty;
            this.category = category;
        }
    }

    public enum Region {
        US, EU
    }

    public interface TaxPolicy {

    }

    // Regla existente: un único impuesto plano por región; los libros y la comida están exentos en la UE
    public static double calculateTotal(List<CartItem> cart, Region region) {
        double subtotal = cart.stream()
                .mapToDouble(it -> it.price * it.qty)
                .sum();

        double tax = 0;
        if (region == Region.US) {
            tax = subtotal * 0.07; // 7% plano
        } else if (region == Region.EU) {
            // exenciones ingenuas en línea
            double taxable = cart.stream()
                    .filter(it -> !"books".equals(it.category) && !"food".equals(it.category))
                    .mapToDouble(it -> it.price * it.qty)
                    .sum();
            tax = taxable * 0.2; // 20% plano solo sobre los ítems gravables
        }

        return roundCurrency(subtotal + tax);
    }

    public static double roundCurrency(double amount) {
        return Math.round(amount * 100.0) / 100.0;
    }

    // Uso de ejemplo, mantenido simple para estudiantes
    public static double demoSprout() {
        List<CartItem> cart = List.of(
                new CartItem("p1", 10, 2, "general"),
                new CartItem("b1", 20, 1, "books")
        );
        return calculateTotal(cart, Region.EU);
    }
}
